Add basic unit tests

0x6e opened the pull request:
tst_InstantiateTypes simply iterates a directory of QML files, each containing a single declarative widget and checks that there were no errors creating the component. This test uses the QML Plugin.

tst_Layouts compares the widget hierarchy of a traditionally created widget (.ui file) against a declarative widget hierarchy. Each test checks a specific QLayout. This test uses the DeclarativeWidgetsDocument.

0x6e commented on tests/auto/layouts/tst_layouts.cpp:
```diff
@@ -51,27 +57,28 @@ tst_Layouts::tst_Layouts()
-    declarativeWidget->show();
+void tst_Layouts::vBoxLayout()
+{
+    testLayouts( QWidgetPtr(new VBoxLayoutWidget())
```
Perform every explicit resource allocation on its own statement. This is unsafe. (Duplicates).

krake-kdab commented on extensionplugin/extensionplugin_plugin.cpp:
```diff
@@ -97,8 +97,21 @@ void ExtensionpluginPlugin::registerTypes(const char *uri)
   qmlRegisterType<QAbstractItemModel>();
+  qmlRegisterType<QItemSelectionModel>();
+
+  // ucreatable gui
```
typo: "uncreatable"

krake-kdab commented on tests/auto/instantiatetypes/qml/uncreatable/AbstractItemModel.qml:
```diff
@@ -0,0 +1,24 @@
+
+import QtWidgets 1.0
+
+AbstractItemModel {
```
hmm, does that one even work?
A QAbstractItemModel cannot be instantiated as it is abstract, i.e. has missing pure virtual methods

0x6e replied:
`tst_InstantiateTypes::uncreatableTypes` verifies that uncreatable types cannot be created.

krake-kdab replied:
ah, ok, makes sense

krake-kdab commented on tests/auto/instantiatetypes/qml/uncreatable/HeaderView.qml:
```diff
@@ -0,0 +1,24 @@
+
+import QtWidgets 1.0
+
+HeaderView {
```
Isn't HeaderView registered as an "uncreatable" type?

krake-kdab commented on tests/auto/instantiatetypes/qml/uncreatable/ItemSelectionModel.qml:
```diff
@@ -0,0 +1,24 @@
+
+import QtWidgets 1.0
+
+ItemSelectionModel {
```
also uncreatable?

krake-kdab commented on tests/auto/instantiatetypes/qml/uncreatable/TextDocument.qml:
```diff
@@ -0,0 +1,24 @@
+
+import QtWidgets 1.0
+
+TextDocument {
```
isn't that an uncreatable type?

krake-kdab commented on tests/auto/instantiatetypes/tst_instantiatetypes.cpp:
```diff
@@ -0,0 +1,88 @@
+    , m_qmlEngine(new QQmlEngine(this))
+{
+    // Add extensionplugin import path
+    m_qmlEngine->addImportPath(QStringLiteral("%1/../../../qml").arg(QCoreApplication::applicationDirPath()));
```
that might be better in the initTestCase() slot, see http://doc.qt.io/qt-5/qtest-overview.html#creating-a-test
Then the relative path can be tested and fail

krake-kdab commented on tests/auto/instantiatetypes/tst_instantiatetypes.cpp:
```diff
@@ -0,0 +1,88 @@
+{
+    QDirIterator iterator(QStringLiteral(":/qml/creatable/"), QDirIterator::Subdirectories);
+    while (iterator.hasNext())
+    {
```
lets go with the Qt coding style for newly added code: https://wiki.qt.io/Qt_Coding_Style
i.e. in this case opening brace on the line of the while condition

krake-kdab commented on tests/auto/instantiatetypes/tst_instantiatetypes.cpp:
```diff
@@ -0,0 +1,88 @@
+        {
+            for (auto error : component.errors())
+            {
+                QWARN(error.toString().toLatin1().data());
```
Instead of toLatin1().data() you can use qPrintable(error.toString())

krake-kdab commented on tests/auto/instantiatetypes/tst_instantiatetypes.cpp:
```diff
@@ -0,0 +1,88 @@
+
+        QQmlComponent component(m_qmlEngine);
+        component.loadUrl(url);
+        if (component.isError())
```
should probably also call QQmlComponent::create() to verify the actual instantiability

krake-kdab commented on tests/auto/layouts/tst_layouts.cpp:
```diff
@@ -0,0 +1,230 @@
+    QVERIFY(uiWidget != nullptr);
+    QVERIFY(declarativeDocument != nullptr);
+
+    QWidgetPtr declarativeWidget(declarativeDocument->create<QWidget>());
```
I would say we switch this to use QQmlComponent for loading and creating the DeclarativeWidget tree.
Otherwise we will have to change this all over again when we deprecate DeclarativeWidgetsDocument in favor of the plugin

0x6e replied:
I agree, done.

0x6e commented:
I have rebased onto Sergio's changes, updated the code style for the unit tests, switched to QQmlComponent instead of DeclarativeWidgetsDocument and refactored tst_Layouts to be a data driven test (to be able to use QEXPECT_FAIL). I ended up not being able to use QEXPECT_FAIL in the way I wanted, but left the changes anyway.

tst_Layouts fails because we cannot apply custom SizePolicy from QML yet. I've added a JIRA task to add SizePolicy as an attached property.

krake-kdab commented on tests/auto/instantiatetypes/tst_instantiatetypes.cpp:
```diff
@@ -42,15 +43,20 @@ tst_InstantiateTypes::tst_InstantiateTypes()
+{
     // Add extensionplugin import path
-    m_qmlEngine->addImportPath(QStringLiteral("%1/../../../qml").arg(QCoreApplication::applicationDirPath()));
+    m_qmlEngine->addImportPath(QStringLiteral("%1/../../../qml")
```
I'd say add a QVERIFY that the path actually exists before passing it to addImportPath.

0x6e replied:
Done.
